\section{Defining regions of interest}
\label{Bioinf_methods: EMT5000 library}

A list of genes known to be involved in the regulation of epithelial-to-mesenchymal transition (EMT) was taken from DeCraene et al. \cite{DeCraene:2013kb}. Regions of interest within the promoter of these genes were identified manually by inspection of chromatin marks in UCSC genome browser (\textbf{table \ref{tab:EMT5000-targets}}). 

\begin{table}[H]
\begin{center}
\begin{footnotesize}
\begin{tabular}{llll}
\hline
Chromosome & Start & Stop & Gene \\
\hline
chr1 &	170626538 &	170637878 &	PRRX1 \\
chr2	& 145272896 & 145282545 & ZEB2 \\
chr2 &	145310788 & 145311630 & ZEB2 \\
chr6 &	166578775 & 166584033 & Brachyury (T gene) \\
chr6	& 166586466 & 166588249 & Brachyury (T gene) \\
chr7	& 19155427	 & 19162115 & TWIST1 \\
chr8	& 49831094	 & 49838789 & SLUG \\
chr10 & 31549929 & 31552360 & ZEB1 \\
chr10 & 31603262 & 31611019 & ZEB1 \\
chr14 & 61113258 & 61126351 & SIX1 \\
chr14 & 95235188 & 95236645 & GSC (Goosecoid) \\
chr16 & 68765472 & 68768900 & Cdh1 \\
chr16 & 68770501 & 68779468 & Cdh1 \\
chr16 & 86596822 & 86601033 & FOXC2 \\
chr18 & 25616470 & 25616815 & Cdh2 \\
chr18 & 25753143 & 25759002 & Cdh2 \\
chr18 & 25763319 & 25764152 & Cdh2 \\
chr18 & 25783828 & 25784775 & Cdh2 \\
chr18 & 52966479 & 52970132 & TCF4 \\
chr18 & 52983584 & 52991765 & TCF4 \\
chr18 & 52994740 & 52997201	& TCF4 \\
chr18 & 53067559 & 53071402	& TCF4 \\
chr18 & 53072594 & 53073776	& TCF4 \\
chr18 & 53087724 & 53090359	& TCF4 \\
chr18 & 53176467 & 53178784	& TCF4 \\
chr18 & 53252816 & 53257791	& TCF4 \\
chr18 & 53259747 & 53260381	& TCF4 \\
chr18 & 53301547 & 53303603	& TCF4 \\
chr19 & 1631468	 & 1633670	& E47/TCF3 \\
chr19 & 1646514	 & 1653855	& E47/TCF3 \\
chr19 & 1655335	 & 1656204	& E47/TCF3 \\
chr19 & 1660918	 & 1661677	& E47/TCF3  \\
chr20 & 48592707 & 48600991 & SNAIL1  \\
chrX & 56258091 & 56260688 &KLF8 \\
\hline
\end{tabular}
\caption{Regions of interest around the promoter regions of 15 genes known to be involved in the regulation of EMT \cite{DeCraene:2013kb}}
\label{tab:EMT5000-targets}
\end{footnotesize}
\end{center}
\end{table}


The genomic regions listed in \textbf{table \ref{tab:EMT5000-targets}} where chosen as target sites for the design of gRNAs  and saved in the file \verb|EMT_genepromoter_comprehensive.gff|. All gRNAs falling into these regions were then found using the Bedtools \cite{Quinlan:2010km} intersect function on the file containing all gRNAs with an NGG PAM found in the genome (see section \ref{sec:GN20GG in genome} for how this file was generated).

\begin{lstlisting}
intersectBed -a GN20GG_masked_autoXY.gff -b EMT_genepromoter_comprehensive.gff
-f 1 -wa -wb > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive.gff
\end{lstlisting}

Guide RNAs were then aligned to the genome in order to identify those that map uniquely. Before alignment, files had to be converted into the correct format as follows:

\begin{lstlisting}
##for alignment need to first convert file into fasta format
##separate + and - strand
cut -f 1,2,3,4 GN20GG_masked_autoXY_EMT_genepromoter_comprehensive.gff | grep '+' | awk '{print "chr" $1 ":" ($2-1) "-" $3}' > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_PLUS.2bit;

cut -f 1,2,3,4 GN20GG_masked_autoXY_EMT_genepromoter_comprehensive.gff | grep -v '+' | awk '{print "chr" $1 ":" $2 "-" $3}' > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS.2bit;

##convert to FASTA format using twoBitToFa
twoBitToFa /path2/human/GRCh37/hg19.2bit GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_PLUS.fa seqList=GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_PLUS.2bit;

twoBitToFa  /path2/human/GRCh37/hg19.2bit GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS.fa seqList=GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS.2bit;

##create reverse complement of guide RNAs on the minus
##strand using the script reverse_complement_fasta.py

python reverse_complement_fasta.py GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS.fa > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS_RC.fa;

##combine the files for the + and - strand
cat GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_PLUS.fa GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_MINUS_RC.fa > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive.fa;
\end{lstlisting}

The PAM sequence was removed using \verb|trimming.py| and alignment to the genome without the PAM (i.e. as GN19 instead of GN20GG).

\begin{lstlisting}
bwa aln -n 0 -o 0 -l 10 -N -I ~/path_to/GRCh37/human_GRCh37.tmp
GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_noPAM.fa >
GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_noPAM.sai;

bwa samse -n 10000 ~/path_to/GRCh37/human_GRCh37.tmp   
GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_noPAM.sai 
GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_noPAM.fa > 
GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_noPAM.sam;

samtools view -S -q1  GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_noPAM.sam | cut -f  1 | awk -F ':' '{print $1 "\t" $2}' | awk -F '-' '{print $1 "\t" $2 "\t" $3}' | sortBed | uniq > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_noPAM_unique.bed
\end{lstlisting}

This file contains 5086 sequences, i.e. 5086 gRNA sequences of the form GN19 that align uniquely to the genome and are followed by an NGG PAM and are found in the regions of interest.

For cloning into the gRNA vector pgRNA-pLKO.1 (see Methods) by Gibson cloning, vector-derived sequences were attached to either end of the gRNA sequence. To this end the FASTA files for the final set of 5086 unique gRNAs were retrieved and sequences appended as follow:

\begin{lstlisting}
#split according to + and - strand 

grep '+' GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand.bed | awk '{print $1 ":" $2 "-" $3}' > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_PLUS.2bit

grep -v '+' GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand.bed | awk '{print $1 ":" $2 "-" $3}' > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_MINUS.2bit

#convert to FASTA format and reverse-complement sequences on the 
#minus strand

twoBitToFa /path2/human/GRCh37/hg19.2bit GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_PLUS.fa seqList=GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_PLUS.2bit;

twoBitToFa /path2/human/GRCh37/hg19.2bit GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_MINUS.fa seqList=GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_MINUS.2bit

python reverse_complement_fasta.py GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_MINUS.fa > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_ noPAM_unique_strand_MINUS_RC.fa

#merge the two files

cat GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_
noPAM_unique_strand_PLUS.fa GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_
noPAM_unique_strand_MINUS_RC.fa > GN20GG_masked_autoXY_EMT_genepromoter_comprehensive_complete_
noPAM_unique_strand.fa

# use trimming.py to remove PAM (because coordinates extracted from SAM file contained PAM, while alignment was run without PAM) and paste the vector sequences for Gibson cloning on either side of the gRNA sequence

sed 'n; s/$/GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCT/' GN20GG_masked_autoXY_EMT_genepromoter_ comprehensive_complete_noPAM_unique.fa | sed 'n; s/^/TCTTGTGGAAAGGACGAAACACC/g' | paste - - | awk '{print $2 "\t" $1}' > EMT_guides_Custom_Array.txt
\end{lstlisting}

A pool of sequences of the form 5'-TCTTGTGGAAAGGACGAAACACC-GN19-GTTTTAGAGCT AGAAATAGCAAGTTAAAATAAGGCT-3', where GN19 donates the 5086 different guide sequences was then ordered from Custom Array Inc. 

